File: javascript/234.Palindrome-Linked-List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  let current = slow;
  let prev = null;

  while (current) {
    const next = current.next;
    current.next = prev;
    prev = current;
    current = next;
  }

  slow = prev;

  let node1 = head;
  let node2 = slow;

  while (node2) {
    if (node1.val !== node2.val) {
      return false;
    }

    node1 = node1.next;
    node2 = node2.next;
  }

  return true;
};

File: python/234.Palindrome-Linked-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # find middle
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # reverse 2nd half
        prev = None
        node = slow
        
        while node:
            nxt = node.next
            node.next = prev
            prev = node
            node = nxt

        # loop and find
        head1 = head
        head2 = prev

        while head2:
            if head1.val != head2.val:
                return False

            head1 = head1.next
            head2 = head2.next

        return True